### DISCUSSION TITLE
Adjust `switch-to-buffer` to show only buffers related to the current activity

### DISCUSSION THREAD

**Discussion by elzorrorebelde** [alphapapa/activities.el#63] [Ideas]
I'm reading the source code (not that I'm that versed in Emacs lisp) and I see no function that takes charge of modifying `switch-to-buffer` behavior.

Ideally, if I'm inside an activity, I would like to see just the buffers related to that activity, omitting specials buffers and buffers from other activities.

**Comment by alphapapa** [comment]
Well, I was waiting for someone to bring this idea up.  :)

The question is, how do we know which buffers are related to an activity?

We could try to use this list of buffers: https://github.com/alphapapa/activities.el/blob/83db6b55476d5fb49e21e52a05643178f6b8b992/activities-tabs.el#L107-L113  But if any unrelated buffer is shown in a window, it would end up in the list.

**Comment by alphapapa** [comment]
Ok, I've implemented this for `activities-tabs-mode`, and it seems to work well.  It can probably be implemented for non-tabs-mode by using frames' buffer lists.  Pushed to `master`.  Please let me know how it works for you.

**Comment by hpfr** [comment]
This is nice! Got me to try the package out for the first time. I have to admit I am somewhat confused on how people are able to use the tab bar and activities-esque packages without some mechanism of grouping visited buffers beyond those currently visible in windows. I typically only have two windows open, so resuming an activity would amount to restoring two buffers. Genuinely curious to hear how people use activities who only care about the visible state. To me that approach doesn’t provide enough utility (vs., say, `project-other-tab-command` or `other-tab-prefix` followed by something like `consult-dir`) to warrant the additional bindings to remember and state to manage. Maybe other people use way more windows? I guess you do get the named tabs as well.

But I digress. Now that this `activities-buffer-list` tab parameter exists, would it be it possible to optionally serialize it as part of suspend and resume, so that opening an activity brings back the window configuration and all serializable buffers, not just those visible in a window? I suppose a command to remove buffers from `activities-buffer-list` without killing them entirely may then be useful as well. For me, I think that would be sufficient functionality to switch from `persp-mode`. Thanks for your work!

**Comment by alphapapa** [comment]
> I have to admit I am somewhat confused on how people are able to use the tab bar and activities-esque packages without some mechanism of grouping visited buffers beyond those currently visible in windows. 

Well, as you might imagine, I use https://github.com/alphapapa/bufler.el for that.  I use `bufler-switch-buffer` on `C-x b`.  I also use `C-x p f` a lot.  And now sometimes `C-x C-a b`.

> I typically only have two windows open, so resuming an activity would amount to restoring two buffers. 

Most of the time I do also.  However I find it helpful to have various "project overview" window configurations, e.g. showing magit-status, the changelog section of the readme, and the notes file, all in one view.  Or for professional work, to have an Org file in one window, plus 3 Org QL View buffers showing various agenda-like queries of tasks and projects in that Org file.  

It would be very tedious to set up such window configurations manually each time--so tedious that it would be impractical.  And that impracticality, IMO, tends to limit users' imagination about what could be done--since it's practically impossible, they don't even consider it.  It's similar to the difficulty of expressing an idea for which there is no word in one's language.

> To me that approach doesn’t provide enough utility (vs., say, project-other-tab-command or other-tab-prefix followed by something like consult-dir) to warrant the additional bindings to remember and state to manage. Maybe other people use way more windows? I guess you do get the named tabs as well.

> But I digress. Now that this activities-buffer-list tab parameter exists, would it be it possible to optionally serialize it as part of suspend and resume, so that opening an activity brings back the window configuration and all serializable buffers, not just those visible in a window? I suppose a command to remove buffers from activities-buffer-list without killing them entirely may then be useful as well. For me, I think that would be sufficient functionality to switch from persp-mode. Thanks for your work!

@hpfr It might be helpful to read this comment about the design principles: https://github.com/alphapapa/activities.el/issues/46#issuecomment-1973862901

I'm not trying to get anyone to switch from `persp-mode` or anything else.  I'm only trying to offer "lightly opinionated" tools that can be used to enhance one's workflows in Emacs.  But I do think that `activities` (and `burly` before it) offer unique tools that haven't existed in Emacs before, and I think that many more users would find them useful if they understood them.

Thanks for your comment.

**Comment by hpfr** [comment]
Thanks for the info on your workflow! Good point about blind spots for workflows that are too tedious by default. For me, I don’t think the knowledge that I will lose a complex window configuration is keeping me from making them—whenever I have three or more windows open, I tend to notice I’m not really using one or more and would rather have more content visible from the rest. I tend to use a large font size as well, so I definitely can see how this approach would be more of a personal quirk than I thought!

As for persisting the buffer list, your linked issue seems more focused on discussion of scope than this one, so I replied there to try to centralize that and limit this issue to feedback on the current buffer switching implementation.

**Comment by elzorrorebelde** [comment]
> I'm not trying to get anyone to switch from `persp-mode` or anything else.

well, now I see that I was mistaken with the scope of this package, my bad.

**Comment by alphapapa** [comment]
> > I'm not trying to get anyone to switch from `persp-mode` or anything else.
> 
> well, now I see that I was mistaken with the scope of this package, my bad.

*shrug*  If `activities` can fulfill similar needs, I'm not necessarily opposed to that, and I'm open to discussing how.  But I'm not necessarily trying to do exactly what `persp` or other "workspace" packages do, either.  There aren't exactly any rules here.  :)

**Comment by alphapapa** [comment]
FWIW, this is what I'm using now, and works well for me:
```el
(defun ap/switch-to-buffer ()
  "Switch to buffer offered from various sources.
If an activity is current, use `activities-switch-buffer';
otherwise use `bufler-switch-buffer'."
  (interactive)
  (cond ((or (equal '(16) current-prefix-arg)
             (not (activities-current)))
         (call-interactively #'bufler-switch-buffer))
        (t (call-interactively #'activities-switch-buffer))))
```

**Comment by jdtsmith** [comment]
> it can probably be implemented for non-tabs-mode by using frames' buffer lists.

Please do, for us frame-users!

**Comment by julian-hoch** [comment]
This is cool stuff, but how do you remove buffers from the activity? I use tabspaces now and it has ways to remove buffers from a tabspace. Perhaps I am missing something - I am still trying to figure this out. I am wondering if in the future activities and tabspaces could work together, since they seem to complement each other rather than compete.

**Reply by alphapapa** [reply]
It's not intended that the user should need to manually control which buffers are associated with an activity.  (I find that kind of micro-managing to be tedious.  I want the computer to do that for me.)  When an activity is opened, the buffers displayed in its frame/tab are associated with it until the activity is closed.

I don't know how tabspaces works, so I can't offer help there.  (I can't keep in my mind how all the various *spaces packages work; that's one reason I wrote this one.)

Anyway, if one really wanted, one could write a command to remove a buffer from an activity's buffer list.  But the idea is that it shouldn't be necessary; it's basically garbage-collected automatically.

**Comment by julian-hoch** [comment]
Thanks for taking the time to reply, I appreciate it!
If the idea is that buffers are removed automatically, then I am curious to see how that can be used currectly - from the little I have seen so far, if I accidentally add a buffer to an activity, there is no way to undo that mistake.

**Reply by jdtsmith** [reply]
Just remove the buffer from the current tab or window, and activities will remove it from the list automatically. 

But there's a better way.  There are two activity states for each activity: default and last.  You initialize with the default activity state, and then can revert to that default state whenever you "pollute" your activity.  This simple trait is the key super-power of activities, IMO.   As your needs change, you can redefine the default state.

**Reply by alphapapa** [reply]
> If the idea is that buffers are removed automatically, then I am curious to see how that can be used currectly - from the little I have seen so far, if I accidentally add a buffer to an activity, there is no way to undo that mistake.

The key is this:

>  When an activity is opened, the buffers displayed in its frame/tab are associated with it until the activity is closed.

In other words, the associated list of buffers is not preserved when an activity is closed.

So, my preference is to not worry if a few stray buffers end up in an activity's associated-buffers list during a session.  They don't significantly interfere with my switching to one of the other associated buffers--compared to having to choose between every buffer open in Emacs, that's nothing.  Then, when I close the activity and reopen it in the future, those stray buffers will be gone.

JD may also be correct that reverting to the activity's default state resets the buffer list--I don't recall at the moment.

Anyway, if you really want to manually manage the list of buffers, a simple command could be written to remove one.

**Reply by jdtsmith** [reply]
My apologies, I may have been responding to an unposed question, since I don't use (emacs) tabs.  Now I gather you are referring to a separate non-visible list of "buffers I _may_ want to show in this activity" which is stored as a tab parameter (though not serialized/restored).

I'd find maintaining such a list somewhat bothersome, because I very commonly "cheat" and quickly switch to an unrelated or slightly related buffer while in an active activity (knowing `revert` is always there to save me).  And of course some useful buffers have generic names (`*Help*`), so storing them permanently is of questionable value.  I suspect this list would quickly come to resemble my browser history ;).  I don't see at present how reverting to `default` resets the list; it seems to be a completely separate out-of-band store of "things I've visited in this tab" — something which, btw, could just as easily exist entirely outside of activities.

It would be straightforward to "bless" this list by bringing it into the activity struct itself (`extra`?), but I think it would take some real thought to make `extra` useful, at least for me.  I could possibly see "opting in" and marking a currently visible buffer and/or associated file as "of interest" by placing (a bookmark to) it on `extra` for the activity.   Then an overloaded `switch-buffer` could consult that list, and/or a consult category could be defined for it for `consult-buffer`.  But it's a bit cumbersome.

One automatic and somewhat intuitive way to store `extra` buffers/files without incurring too much "pollution": every time you `redefine` an activity, keep track of all the buffers/files that got replaced on `extra` as "possibly of interest".  The reason this might work is (in my usage at least) `redefine` is a serious, planful step: no garbage buffers are likely visible when I `redefine`.  

_This_ list I suspect would be useful and high-signal enough for me to reach for at times.

**Reply by alphapapa** [reply]
@jdtsmith I chose not to store the buffer list in the struct because I wanted it to be ephemeral; so it's stored as either a frame or tab parameter, depending on whether `activities-tabs-mode` is active.  

Storing it in the struct would also mean that code would have to be written to (de)serialize those buffers in the struct; and then we'd have to decide whether those buffers should actually be recreated with the activity, or just offered for selection if already opened (which would be even more complicated).

I can imagine how it could be useful to recreate buffers other than the ones displayed in an activity's window configuration, so I'm willing to discuss those ideas.  But given the myriad other ways to recreate such buffers "normally" (e.g. finding files with project commands, bookmarks, dired, xref; commands that open special buffers; defining other activities; etc), that seems of somewhat questionable usefulness and priority.  I'm also not sure if the additional complexity would be worth it.

**Reply by jdtsmith** [reply]
Yeah, I fully agree about all the other ways to manage lists of buffers.  If this were to be pursued further, I think it should be _very_ activities-centric.  Perhaps a curated list, the test for which is "did I ever include this in my default for this activity"?   Otherwise there's probably too much overlap with other tools in this space.   

In terms of mechanics, it's easy to save bookmarks to such buffers/files, yes?  I.e. offload all the difficulties to bookmark.

**Reply by alphapapa** [reply]
Bookmarks is how the whole thing works internally; the tedium comes from manipulating the deeply nested data structures.
